### PR TITLE
RUST-713 Fix underflow on BSON array and binary deserialization

### LINKED ISSUE #243: OOM on huge allocation on malformed input (found through fuzzing)
Minimised base64 input: `CgAAAAQAAAAAAA==`

Stack trace:

```
==3694748==ERROR: AddressSanitizer: requested allocation size 0xfffffffffffffffc (0x800 after adjustments for alignment, red zones etc.) exceeds maximum supported size of 0x10000000000 (thread T0)
    #0 0x55e5687daf12 in calloc /rustc/llvm/src/llvm-project/compiler-rt/lib/asan/asan_malloc_linux.cpp:154:3
    #1 0x55e56881c330 in bson::de::ensure_read_exactly::hf5185d965cc60fad (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/deserialize+0x21f330)
    #2 0x55e568819981 in bson::de::deserialize_array::ha4650ca0cb324222 (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/deserialize+0x21c981)
    #3 0x55e56881e3da in bson::de::deserialize_bson_kvp::hec5fdee461eb2b1b (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/deserialize+0x2213da)
    #4 0x55e568828b98 in bson::document::Document::decode::_$u7b$$u7b$closure$u7d$$u7d$::hf6f570c9035f9cf2 (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/deserialize+0x22bb98)
    #5 0x55e56881bcab in bson::de::ensure_read_exactly::h6a68518f15fd1165 (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/deserialize+0x21ecab)
    #6 0x55e568828653 in bson::document::Document::decode::hab1f24b5dff83160 (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/deserialize+0x22b653)
    #7 0x55e56884eb2c in rust_fuzzer_test_input (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/deserialize+0x251b2c)
    #8 0x55e568c35280 in __rust_try (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/deserialize+0x638280)
    #9 0x55e568c34edf in LLVMFuzzerTestOneInput (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/deserialize+0x637edf)
    #10 0x55e568c2c104 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/deserialize+0x62f104)
    #11 0x55e568c2c882 in fuzzer::Fuzzer::MinimizeCrashLoop(std::vector<unsigned char, fuzzer::fuzzer_allocator<unsigned char> > const&) (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/deserialize+0x62f882)
    #12 0x55e568c1b0ef in fuzzer::MinimizeCrashInputInternalStep(fuzzer::Fuzzer*, fuzzer::InputCorpus*) (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/deserialize+0x61e0ef)
    #13 0x55e568c2500d in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/deserialize+0x62800d)
    #14 0x55e56875f142 in main (/home/jess/.cache/cargo/target/x86_64-unknown-linux-gnu/release/deserialize+0x162142)
    #15 0x7f4586739b24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)
```

### PR DESCRIPTION
**Pull request by gperinazzo**
Closes #243

### CODE REVIEW

**Comment by agolin95** [comment]
Tracked in https://jira.mongodb.org/browse/RUST-713

**Review by patrickfreed — COMMENTED** [review]
Thank you @gperinazzo for submitting this patch, and also thank you @5225225 for identifying reproducible crash cases!

The fix looks good to me, but could we add the repros as test cases to ensure we don't accidentally introduce a regression in the future?

**Comment by gperinazzo** [comment]
Hey @patrickfreed, I added tests for #243 in the same format as the ones for #64. Let me know if you would like any more changes.

File: src/tests/modules/serializer_deserializer.rs [hunk 1]
```diff
@@ -540,3 +540,20 @@ fn test_serialize_deserialize_document() {
     let bad_point: Result<Point, crate::de::Error> = from_document(bad_point);
     assert!(bad_point.is_err());
 }
+
+#[test]
+fn test_deserialize_invalid_array_length_issue243() {
+    let _guard = LOCK.run_concurrently();
+    let buffer = b"\n\x00\x00\x00\x04\x00\x00\x00\x00\x00";
+    let _ = Document::from_reader(&mut std::io::Cursor::new(buffer));
```
Comment (patrickfreed) [inline src/tests/modules/serializer_deserializer.rs:548]: Here and below, can we add an `expect_err` assertion to ensure we're rejecting the invalid inputs?

**Review by patrickfreed — COMMENTED** [review]
Thanks! Could we actually drop the issue number from the test name and instead add a link to the Jira ticket (https://jira.mongodb.org/browse/RUST-713) in a docstring for the test instead? I think that way it'll be easier to follow for others seeing it for the first time (or for me in the future :slightly_smiling_face:). I think that test name format was used from before MongoDB started officially maintaining this library.

**Comment by gperinazzo** [comment]
I've replaced the issue in the function name with a link to the Jira ticket as a docstring for the test and added the `expect_err`.

**Comment by patrickfreed** [comment]
Code changes LGTM! I authorized the CI run and in the meantime am tagging in the other members of the team as reviewers.

**Review by saghm — APPROVED** [review]

**Review by isabelatkinson — APPROVED** [review]
LGTM other than the lint error

**Review by patrickfreed — APPROVED** [review]
I went ahead and fixed the lint for you, if that's okay. Once the CI comes back green I'll merge it in, thanks again for your contribution!

**Pull request merged**